### DISCUSSION TITLE
Template API: Idea for improving block dependencies on other blocks

### DISCUSSION THREAD

**Discussion by mattsherman** [woocommerce/woocommerce#40026] [WooCommerce New Product Editor] [answered]
## Background

In #39915, we registered `SimpleProductTemplate` and added an action, `woocommerce_block_template_register`, that is fired after a template is registered with the `BlockTemplateRegistry`.

One main potential usage of this action was provide extensions with a well-defined point to make template modifications (such as block addition or removal).

## Issue

As @joshuatf pointed out in the review (https://github.com/woocommerce/woocommerce/pull/39915#discussion_r1306476814), using this action to make template modifications implies that any block dependencies (such as a group/section that you are adding a block to; or a block that you are removing), have been added to the template at the point when this action is fired.

However, nothing enforces that all blocks are added to the template before this action is called. We could do this for the core blocks in the template (we are actually already doing so). But, extensions have no way to do so, and we run into an issue of filter priority if extensions are trying to make modifications that are dependent on modifications made by another extension.

Things could get messy and complicated quickly! And, ideally we don't have any special handling or different behavior for 'core' vs extension blocks in templates.

### `woocommerce_block_template_after_add_block` action

Building off of a suggestion by @joshuatf (https://github.com/woocommerce/woocommerce/pull/39915#discussion_r1307732026), and one that was also initially internally discussed early on (I can't find a reference; it may have been an in-person/video discussion), let's add an action that will be called after a block is added to a template.

This action can be used to:

* Add a dependent block
* Set up conditional visibility of a block (future API capability)

Example code that adds a block dependent on an existing block:

```php
if ( ! function_exists( 'YOUR_PREFIX_block_template_after_add_block' ) ) {
	function YOUR_PREFIX_block_template_after_add_block( BlockInterface $block ) {
		$block_id = $block->get_id();

		if ( 'product-summary' === $block_id ) {
			// Add a toggle after the summary field
			$parent = $block->get_parent();

			if ( method_exists( $parent, 'add_block' ) ) {
				$parent->add_block(
					[
						'id'         => 'your-prefix-summary-show-toggle',
						'blockName'  => 'woocommerce/product-toggle-field',
						'order'	     => $block->get_order() + 5,
						'attributes' => [
							'label'    => __( 'Show summary', 'woocommerce' ),
						],
					]
				);
			}
		}
	}	

	add_action( 'woocommerce_block_template_after_add_block', 'YOUR_PREFIX_block_template_after_add_block' );
}
```

Since this action is called after a block has been added to a template, extensions can be assured that the block they are dependent on (such as the group/section they want to add their block to) is in the template.

Well... except for one small issue... **what if the block they are dependent on is later removed via `remove_block()`** (added in #39900)? 

We could add a similar action that was called after a block was removed, but that gets messy to handle. Extensions would make their changes after their dependency was added and then have to roll it back after the dependency was removed. Surely we should avoid such complexities.

It would be so much easier if we just didn't let blocks get removed! I'm serious... let's not let blocks get removed after they have been added to a template. Rather, let's provide a means to prevent blocks from being added in the first place...

### `woocommerce_block_template_before_add_block` filter

Let's add a filter that will be called before a block is added to a template. By returning `false`, the block would not be added.

This filter can be used to:

* Prevent a block from being added, by returning `false`
* Replace a block, by returning `false` and adding another block to the same parent container, with the same `order`

Example code that prevents an existing block from being added to the template:

```php
if ( ! function_exists( 'YOUR_PREFIX_block_template_before_add_block' ) ) {
	function YOUR_PREFIX_block_template_before_add_block( bool $should_add_block, BlockInterface $block ) {
		$block_id = $block->get_id();

		if ( 'product-summary' === $block_id ) {
			// Remove the summary field
			$should_add_block = false;
		}

		return $should_add_block;
	}

	add_filter( 'woocommerce_block_template_before_add_block', 'YOUR_PREFIX_block_template_before_add_block', 10, 2 );
}

```

Example code that replaces an existing block in the template:

```php
if ( ! function_exists( 'YOUR_PREFIX_block_template_before_add_block' ) ) {
	function YOUR_PREFIX_block_template_before_add_block( bool $should_add_block, BlockInterface $block ) {
		$block_id = $block->get_id();

		if ( 'product-sale-price' === $block_id ) {
			// Remove the sale price field and replace it with a toggle
			$should_add_block = false;

			$parent = $block->get_parent();

			if ( method_exists( $parent, 'add_block' ) ) {
				$parent->add_block(
					[
						'id'         => 'your-prefix-sale-price-toggle',
						'blockName'  => 'woocommerce/product-toggle-field',
						'order'	     => $block->get_order(),
						'attributes' => [
							'label'    => __( 'On sale', 'woocommerce' ),
						],
					]
				);
			}
		}

		return $should_add_block;
	}

	add_filter( 'woocommerce_block_template_before_add_block', 'YOUR_PREFIX_block_template_before_add_block', 10, 2 );
}
```

We would remove the `remove_block()`/`remove_blocks()` API methods and this filter would be the only means for a block to be "removed" from a template.

## Try PR

You can give this approach a spin by checking out the following try PR:

- https://github.com/woocommerce/woocommerce/pull/40010

## Feedback desired!

Please let me know what you think about this proposed change to the template API. Are there any downsides or details that I haven't considered?

**Comment by joshuatf** [comment]
Thanks for opening up this discussion, @mattsherman!

>We could add a similar action that was called after a block was removed, but that gets messy to handle. Extensions would make their changes after their dependency was added and then have to roll it back after the dependency was removed. Surely we should avoid such complexities.

I think in the majority of cases, the dependent block is not a sibling, but a parent.  For example, a block may add itself to the "Basic details" section.  In these cases, no further action is needed by the plugin extending this section since their block has also been removed by the removal of this section.

In cases where it's a sibling element, I think a `woocommerce_block_template_after_remove_block` hook would not complicate things much to remove.  

```
function remove_my_block( $block, $template ) {
    if ( $block->get_id() === 'sibling-block' ) {
        $template->remove_block( 'my_block' );
    }
}
add_action( 'woocommerce_block_template_after_remove_block', 'remove_my_block', 10, 2 );
```

If we think that's too much, we could always create a wrapper around this logic (something like `$block->append_sibling( ... )`) that handles this boilerplate.  But IMO the above is not overly complex and is more of an edge case in comparison to adding blocks to parent sections.

>It would be so much easier if we just didn't let blocks get removed! I'm serious... let's not let blocks get removed after they have been added to a template. Rather, let's provide a means to prevent blocks from being added in the first place...

I'm not a huge fan of restricting what can be removed from blocks.  Not only does it limit the API itself and create a heavier reliance on hooks, but this opens the door for even more timing issues.  Depending on when a block is added, a filter preventing its addition may or may not have been registered in time.  And since a block can be registered at any point during the lifecycle, we can't provide a reliable hook for when this filter should be added by.

**Reply by mattsherman** [reply]
> I think in the majority of cases, the dependent block is not a sibling, but a parent.

I agree, though I do think the sibling dependency will become popular since it can be used (for good or for bad) to position blocks in relation to siblings without having to hardcode `order` values (I know that is something I have liked doing when testing out the API).

> I'm not a huge fan of restricting what can be removed from blocks. Not only does it limit the API itself

I don't see this proposal as restricting anything. It just changes how a block is "removed". Can you see scenarios where a block would not be able to be removed with this approach?

Another thing I like about the `woocommerce_block_template_before_add_block` approach is that extensions (and core) can know right away when they add their blocks if they are being "removed" (without having to use another hook).

> create a heavier reliance on hooks

I don't see a problem with relying on hooks if they are well documented and easy to understand.

> this opens the door for even more timing issues. Depending on when a block is added, a filter preventing its addition may or may not have been registered in time. And since a block can be registered at any point during the lifecycle, we can't provide a reliable hook for when this filter should be added by.

I don't see how the `woocommerce_block_template_after_remove_block` suggestion solves this. If that filter isn't hooked up before the removal, we would face the same issue.

In order to provide for assurance in the lifecycle, we should introduce lifecycle actions that we support for templates (regardless of how we handle removal, I think these will be useful)...

- `woocommerce_block_template_registry_pre_init` - Extensions would hook up any template modification hooks in here, so they are set before blocks are added to templates, such as...
    - `woocommerce_block_template_after_add_block `
    - `woocommerce_block_template_before_add_block ` (and/or, `woocommerce_block_template_after_remove_block`)
- `woocommerce_block_template_registry_init` - This is when templates should be registered, and blocks added to templates
- `woocommerce_block_template_register` (existing; we should consider renaming to `woocommerce_block_template_after_register` for clarity) - Extensions can be assured that all templates are registered by this point, and add blocks to templates that don't depend on any existing blocks (I don't know if this will ever be the case though, since they will depend on the container at the very least!)

**Reply by joshuatf** [reply]
>I don't see this proposal as restricting anything. It just changes how a block is "removed". Can you see scenarios where a block would not be able to be removed with this approach?

It is restrictive in the sense that we are limiting the timing for when a plugin can prevent a block from being added.

The filter `woocommerce_block_template_before_add_block` occurs when the block is added, which may be early in the page lifecycle.  A plugin that needs information from a data store or something else at a later hook is already too late to prevent this block from being added.

>I don't see a problem with relying on hooks if they are well documented and easy to understand.

I don't see an issue with relying on hooks either, but I would favor a more structured API where possible.  Methods provide clearer intent of what the action being taken is compared to a hook in addition to allowing us more flexibility in the future to change how things operate under the hood.

>I don't see how the woocommerce_block_template_after_remove_block suggestion solves this. If that filter isn't hooked up before the removal, we would face the same issue.

How would we get into a scenario where a filter for removing the block was not able to be added in time, but the filter for adding the block was added in time?

>In order to provide for assurance in the lifecycle, we should introduce lifecycle actions that we support for templates (regardless of how we handle removal, I think these will be useful)...

I'm on board with the mentioned hooks and do think these could be useful.

The one that solves a lot of the timing issues, but we also might want to be exercise some caution is `woocommerce_block_template_registry_init`.  I like this general idea, though we need to make sure this is late enough to allow plugin classes to fully initialize.  Just at a glance, I imagine after `admin_init` and 10+ priority may be a good starting point.

**Comment by mattsherman** [comment, marked as answer]
After discussing this more with @joshuatf (in a 1:1 call), we feel that we should proceed as follows:

- Introduce `woocommerce_block_template_after_add_block` action as proposed
- Keep the `remove_block()` API (as opposed to the proposed `woocommerce_block_template_before_add_block` filter approach)
    - This is more understandable to devs
    - It allows us to determine more explicit intent in API usage, which could be useful for future improvements
- Introduce `woocommerce_block_template_after_remove_block` action to handle any processing that needs to occur after a block is removed
- Introduce `woocommerce_block_template_registry_init` lifecycle action (or, just use the existing `woocommerce_block_template_register` action)
    - This is the point at which it is safe to add blocks to templates (since the above filters will have been hooked up)
- Document that the following actions should be hooked up prior to `woocommerce_block_template_registry_init` (we will make a specific recommendation)
    - `woocommerce_block_template_after_add_block`
    - `woocommerce_block_template_after_remove_block`

I will update the related try branch PR with these changes to prove these out with the same examples that I used previously (see https://gist.github.com/mattsherman/a613580a841eba5b5c402389dbcea88f).
